### ISSUE TITLE
How exactly is syncing with gpodder.net is done?

### ISSUE THREAD

**Issue by doronbehar** [gpodder/gpodder#370]
I'm using [AntennaPod](https://github.com/AntennaPod/AntennaPod) combined with gPodder on my PC to view and listen to podcasts. The support on behalf of AntennaPod for syncing episode states and subscriptions is great. When I'm viewing something on AntennaPod and syncing it to [gpodder.net](https://gpodder.net), I refresh gpodder on the PC and I see episodes read/new state is updated - great.

My question is regarding the reverse sync: After watching and listening to podcasts on the PC and although the configuration option `mygpo.enabled` is set to true, I don't see the AntennaPod device changing the episodes' read/new state. I know I can `Replace list on server with local subscriptions` but Is this the way it is supposed to be? Is the only way to have a full 2-way sync is by pressing this button and potentially losing podcasts on the server and therefor on the other devices?

**Comment by zebastiane** [comment]
I have the same question and didn't find any answer in the documentation.
Episodes not marked as new are not seen as played on AntennaPod.

gPodder can't sync the position because it uses an external player, so maybe that's why syncing with AntennaPod doesn't work?

**Comment by elelay** [comment]
Yes, either not having the total length or playback position <= 0 would prevent gPodder play action to be taken into account by AntennaPod (see the [code](https://github.com/AntennaPod/AntennaPod/blob/f069894b92306c95183aca2eb82496cf94ccf323/core/src/main/java/de/danoeh/antennapod/core/gpoddernet/model/GpodnetEpisodeAction.java#L84)).

If you are on Linux, what is your external player? Have you enabled the 'MPRIS Listener' extension? 
If so, running `gpodder -v` you should see a debug message `Received play action: ...` when you pause playback in your media player. Then gPodder should send play actions with the.

Example log (after letting the audio play ~20s)
```
1549352072.688316 [mpris-listener] INFO: CurrentTrackTracker: Stopped: file:///home/elelay/gPodder/Downloads/99_%20Invisible/Bonus_Episode_Avery_update_on_AoI.mp3: Paused at 19/664 (@1.000000) b'/home/elelay/gPodder/Downloads/99_ Invisible/Bonus_Episode_Avery_update_on_AoI.mp3'
1549352072.688459 [gpodder.gtkui.main] DEBUG: Received play action: file:///home/elelay/gPodder/Downloads/99_%20Invisible/Bonus_Episode_Avery_update_on_AoI.mp3 (3, 19, 664)
1549352072.699764 [gpodder.dbsqlite] DEBUG: Commit.
1549352072.704075 [gpodder.my] DEBUG: Storing full episode playback action
1549352097.852661 [gpodder.config] INFO: Flushing settings to disk
```

**Comment by zebastiane** [comment]
I'm not on Linux but macOS and I'm using VLC.

I don't see the `MPRIS Listener` in my available extensions in the preferences.

The extension is found, so I tried to force it but it didn't work

```
1550142962.816268 [gpodder.extensions] DEBUG: Found extension "mpris-listener" in /Applications/gPodder.app/Contents/Resources/share/gpodder/extensions/mpris-listener.py
…
1550143010.394808 [gpodder.config] DEBUG: extensions.enabled: ['update_feeds_on_startup', 'enqueue_in_mediaplayer'] -> ['update_feeds_on_startup', 'enqueue_in_mediaplayer', 'mpris-listener']
1550143010.395002 [gpodder.extensions] INFO: Extension "mpris-listener" is now enabled
1550143010.395113 [gpodder.extensions] INFO: Not loading "mpris-listener" (only_for = "freedesktop")
```

There is no option to make this work on macOS?

**Comment by dovf** [comment]
The MPRIS Listener is based on D-Bus [1], so you'll need to get D-Bus
working on macOS in order to use it. I'm not at all familiar with macOS, a
quick google search seems to indicate that it may be possible to get D-Bus
working on macOS, though not necessarily easy.

Note that the reason MPRIS Listener requires D-Bus is that it basically
translates between two different D-Bus interfaces (one defined by gPodder
allowing it to get notifications about episode play position, and another
-- MPRIS -- which defines how media players can broadcast notifications
about the media being played, including Start/Stop events and the positions
at which they occur. The MPRIS listener listens to the latter, and
translates into the former so that gPodder can receive the information;
more info at [2]). So once you get D-Bus running, you'll also need to make
sure that both gPodder and your media player are turning on their D-Bus
interfaces, besides forcing the extension to run.

But again, I'm not at all familiar with macOS, there may be a more
"macOS-native" way of dealing with this kind of issue -- though I'm not
aware of gPodder having any builtin support for that...

HTH!
Dov

[1] https://www.freedesktop.org/wiki/Software/dbus/
[2] https://gpodder.github.io/docs/extensions/mprislistener.html

**Comment by zebastiane** [comment]
Thank you for the detailed answer.

I was able to install D-Bus, but unfortunately the MPRIS Listener doesn't load because of `only_for = "freedesktop"`

Seeing this line in the code source I don't think it's not possible de make the extension work on macOS
https://github.com/gpodder/gpodder/blob/e84547ad183c6c333ac9114d7f7e0576e2df60e5/src/gpodder/__init__.py#L110

Another solution would be to mark as read in gpodder, but I only see an option to toggle the new state.

**Comment by elelay** [comment]
I see 3 paths to do what you want, all by writing a gPodder extension:
 1. develop a gPodder extension to add the "Mark as Read" action, maybe using mutagen to set the played position to the file length.
 2. develop a gPodder extension to query VLC using applescript. It [seems to support](https://github.com/videolan/vlc/blob/ba6ad45fbd4f8130139c32901c46cbab875d0871/modules/gui/macosx/os-integration/applescript.h#L40) all the required properties (path, current time, total time), but you would have to poll for changes every x seconds.
3. develop a gPodder extension to subscribe to [MPNowPlayingInfoCenter](https://developer.apple.com/documentation/mediaplayer/mpnowplayinginfocenter) to get vlc's information. It's supported by [PyObjC](https://bitbucket.org/ronaldoussoren/pyobjc/src/3513e3d3ec7df58e09eeafd3467f96ff06dc0982/pyobjc-framework-MediaPlayer/PyObjCTest/test_mpnowplayinginfocenter.py?at=default&fileviewer=file-view-default). I don't know if it really works...

**Comment by dovf** [comment]
> I was able to install D-Bus, but unfortunately the MPRIS Listener doesn't load because of `only_for = "freedesktop"`
> 
> Seeing this line in the code source I don't think it's not possible de make the extension work on macOS

I believe that line is there only *because* D-Bus is by default available only on freedesktop; if you've installed D-Bus, I would try to just remove that line and then see how far you get...
